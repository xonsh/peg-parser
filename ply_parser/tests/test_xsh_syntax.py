"""Tests the xonsh parser."""

import sys
from pathlib import Path

import pytest


def get_cases(path: Path, splitter="# "):
    inp = []
    parts = []

    def value():
        res = ("\n".join(inp), "\n".join(parts))
        inp.clear()
        parts.clear()
        return res

    for line in path.read_text().splitlines():
        if line.startswith(splitter):
            inp.append(line.lstrip(splitter))
        elif not line.strip():
            if parts:
                yield value()
        else:
            parts.append(line)
    if parts:
        yield value()


def glob_data_param(pattern: str):
    for path in Path(__file__).parent.joinpath("data").glob(pattern):
        for idx, (inp, exp) in enumerate(get_cases(path)):
            yield pytest.param(inp, exp, id=f"{path.name}-{idx}")


def yaml_line_items(*names: str):
    for name in names:
        path = Path(__file__).parent.joinpath("data").joinpath(f"{name}.yml")
        import yaml

        with path.open("r") as file:
            data = yaml.safe_load(file)
        for case, lines in data.items():
            for idx, inp in enumerate(lines):
                yield pytest.param(inp, id=f"{path.name}-{case}-{idx}")


@pytest.mark.parametrize("inp", yaml_line_items("exprs", "stmts"))
def test_line_items(inp, unparse, snapshot):
    assert unparse(inp) == snapshot


@pytest.mark.parametrize(("inp", "exp"), glob_data_param("fstring_py312.py"))
@pytest.mark.skipif(sys.version_info < (3, 12), reason="requires python3.12")
def test_py312_fstring(inp, exp, unparse_diff):
    unparse_diff(inp, exp)


@pytest.mark.parametrize(
    "inp",
    [
        'x = "WAKKA"; ${x} = 65',
        'x = "."; $(ls @(None or x))',
        'x = "."; !(ls @(None or x))',
        '$[git commit -am "wakka jawaka" ]',
        '$[git commit -am "flock jawaka milwaka" ]',
        '$[git commit -am "wakka jawaka"]',
        '$[git commit -am "flock jawaka"]',
        '![git commit -am "wakka jawaka" ]',
        '![git commit -am "flock jawaka milwaka" ]',
        '![git commit -am "wakka jawaka"]',
        '![git commit -am "flock jawaka"]',
    ],
)
def test_statements(check_xonsh_ast, inp):
    if not inp.endswith("\n"):
        inp += "\n"
    check_xonsh_ast(inp, mode="exec")


@pytest.mark.parametrize(
    ("inp", "args"),
    [
        ("$(ls)", ["ls"]),
        ("$(ls )", ["ls"]),
        ("$( ls )", ["ls"]),
        ("$( ls)", ["ls"]),
        ("$(ls .)", ["ls", "."]),
        ('$(ls ".")', ["ls", '"."']),
        ("$(ls -l)", ["ls", "-l"]),
        ("$(ls $WAKKA)", ["ls", "wak"]),
        ('$(ls @(None or "."))', ["ls", "."]),
        (
            '$(echo hello | @(lambda a, s=None: "hey!") foo bar baz)',
            ["echo", "hello", "|", "hey!", "foo", "bar", "baz"],
        ),
        (
            "$(echo @(i**2 for i in range(20) ) )",
            [
                "echo",
                0,
                1,
                4,
                9,
                16,
                25,
                36,
                49,
                64,
                81,
                100,
                121,
                144,
                169,
                196,
                225,
                256,
                289,
                324,
                361,
            ],
        ),
        ("$(echo @('a', 7))", ["echo", "a", 7]),
        pytest.param(
            "$(@$(which echo) ls | @(lambda a, s=None: $(@(s.strip()) @(a[1]))) foo -la baz)",
            "",
            marks=pytest.mark.xfail,
        ),
        ("$(ls $(ls))", ["ls", "ls"]),
        ("$(ls $(ls) -l)", ["ls", "ls", "-l"]),
        ("$[ls]", ["ls"]),
        ("![ls]", ["ls"]),
        ("![echo $WAKKA/place]", ["echo", "wak/place"]),
        ("![echo yo==yo]", ["echo", "yo==yo"]),
        ("!(ls | grep wakka)", ["ls", "|", "grep", "wakka"]),
        (
            "!(ls | grep wakka | grep jawaka)",
            ["ls", "|", "grep", "wakka", "|", "grep", "jawaka"],
        ),
        ("!(ls > x.py)", ["ls", ">", "x.py"]),
    ],
)
def test_captured_procs(inp, args, check_xonsh_ast, xsh_proc_method):
    check_xonsh_ast(inp, mode="exec", xenv={"WAKKA": "wak"})
    method = xsh_proc_method(inp[:2])
    method.assert_called_with(*args)


@pytest.mark.parametrize(
    "expr",
    [
        "!(ls)",
        "!(ls )",
        "!( ls)",
        "!( ls )",
        "!(ls .)",
        '!(ls @(None or "."))',
        '!(ls ".")',
        "!(ls $(ls))",
        "!(ls $(ls) -l)",
        "!(ls $WAKKA)",
        "!($LS .)",
    ],
)
def test_bang_procs(expr, check_xonsh_ast):
    check_xonsh_ast(expr, xenv={"LS": "ll", "WAKKA": "wak"})


@pytest.mark.parametrize("p", ["", "p"])
@pytest.mark.parametrize("f", ["", "f"])
@pytest.mark.parametrize("glob_type", ["", "r", "g"])
def test_backtick(p, f, glob_type, check_xonsh_ast):
    check_xonsh_ast(f"print({p}{f}{glob_type}`.*`)", False)


def test_comment_only(check_xonsh_ast):
    check_xonsh_ast("# hello", mode="exec")


@pytest.mark.parametrize(
    "case",
    [
        "![(cat)]",
        "![(cat;)]",
        "![(cd path; ls; cd)]",
        '![(echo "abc"; sleep 1; echo "def")]',
        '![(echo "abc"; sleep 1; echo "def") | grep abc]',
        "![(if True:\n   ls\nelse:\n   echo not true)]",
    ],
)
def test_use_subshell(case, check_xonsh_ast):
    check_xonsh_ast(case)
