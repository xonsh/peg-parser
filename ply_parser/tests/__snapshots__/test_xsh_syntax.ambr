# serializer version: 1
# name: test_line_items[exprs.yml-bg-0]
  "__xonsh__.subproc_captured_stdout(['emacs', 'ugggh'], '&')"
# ---
# name: test_line_items[exprs.yml-bg-1]
  "__xonsh__.subproc_captured_stdout(['emacs', 'ugggh'], '&')"
# ---
# name: test_line_items[exprs.yml-bg-2]
  "__xonsh__.subproc_captured_stdout(['emacs', 'ugggh'], '&')"
# ---
# name: test_line_items[exprs.yml-bg-3]
  "__xonsh__.subproc_captured_stdout(['emacs', 'ugggh'], '&')"
# ---
# name: test_line_items[exprs.yml-combinators-0]
  "__xonsh__.subproc_captured_stdout(['ls'], '|', ['grep', 'wakka'])"
# ---
# name: test_line_items[exprs.yml-combinators-1]
  "__xonsh__.subproc_captured_stdout(['ls'], '|', ['grep', 'wakka'], '|', ['grep', 'jawaka'])"
# ---
# name: test_line_items[exprs.yml-combinators-2]
  "__xonsh__.subproc_captured_hiddenobject(['ls', 'me']) and __xonsh__.subproc_captured_hiddenobject(['grep', 'wakka'])"
# ---
# name: test_line_items[exprs.yml-combinators-3]
  "__xonsh__.subproc_captured_hiddenobject(['ls']) and __xonsh__.subproc_captured_hiddenobject(['grep', 'wakka']) and __xonsh__.subproc_captured_hiddenobject(['grep', 'jawaka'])"
# ---
# name: test_line_items[exprs.yml-combinators-4]
  "__xonsh__.subproc_captured_hiddenobject(['ls']) and __xonsh__.subproc_captured_hiddenobject(['grep', 'wakka'])"
# ---
# name: test_line_items[exprs.yml-combinators-5]
  "__xonsh__.subproc_captured_hiddenobject(['ls']) and __xonsh__.subproc_captured_hiddenobject(['grep', 'wakka']) and __xonsh__.subproc_captured_hiddenobject(['grep', 'jawaka'])"
# ---
# name: test_line_items[exprs.yml-combinators-6]
  "__xonsh__.subproc_captured_hiddenobject(['ls']) or __xonsh__.subproc_captured_hiddenobject(['grep', 'wakka'])"
# ---
# name: test_line_items[exprs.yml-combinators-7]
  "__xonsh__.subproc_captured_hiddenobject(['ls']) or __xonsh__.subproc_captured_hiddenobject(['grep', 'wakka']) or __xonsh__.subproc_captured_hiddenobject(['grep', 'jawaka'])"
# ---
# name: test_line_items[exprs.yml-combinators-8]
  "__xonsh__.subproc_captured_hiddenobject(['ls']) or __xonsh__.subproc_captured_hiddenobject(['grep', 'wakka'])"
# ---
# name: test_line_items[exprs.yml-combinators-9]
  "__xonsh__.subproc_captured_hiddenobject(['ls']) or __xonsh__.subproc_captured_hiddenobject(['grep', 'wakka']) or __xonsh__.subproc_captured_hiddenobject(['grep', 'jawaka'])"
# ---
# name: test_line_items[exprs.yml-env-0]
  "__xonsh__.env['WAKKA']"
# ---
# name: test_line_items[exprs.yml-env-1]
  "__xonsh__.env[None or __xonsh__.env['JAWAKA']]"
# ---
# name: test_line_items[exprs.yml-globs-0]
  "__xonsh__.pathsearch(__xonsh__.regexsearch, 'hello', True, False)"
# ---
# name: test_line_items[exprs.yml-globs-10]
  "__xonsh__.subproc_captured_stdout(['ls'] + __xonsh__.pathsearch(__xonsh__.regexsearch, '[Ff]+i*LE', False, False) + ['-l'])"
# ---
# name: test_line_items[exprs.yml-globs-1]
  "__xonsh__.subproc_captured_stdout(['ls'] + __xonsh__.pathsearch(__xonsh__.regexsearch, '#[Ff]+i*LE', False, False) + ['-l'])"
# ---
# name: test_line_items[exprs.yml-globs-2]
  "__xonsh__.subproc_captured_stdout(['ls'] + __xonsh__.pathsearch(__xonsh__.regexsearch, '[Ff]+i*LE', False, False) + ['-l'])"
# ---
# name: test_line_items[exprs.yml-globs-3]
  "__xonsh__.subproc_captured_stdout(['ls'] + __xonsh__.pathsearch(__xonsh__.regexsearch, '#[Ff]+i*LE', False, False) + ['-l'])"
# ---
# name: test_line_items[exprs.yml-globs-4]
  "__xonsh__.subproc_captured_stdout(['ls'] + __xonsh__.pathsearch(__xonsh__.globsearch, '[Ff]+i*LE', False, False) + ['-l'])"
# ---
# name: test_line_items[exprs.yml-globs-5]
  "__xonsh__.subproc_captured_stdout(['ls'] + __xonsh__.pathsearch(__xonsh__.globsearch, '#[Ff]+i*LE', False, False) + ['-l'])"
# ---
# name: test_line_items[exprs.yml-globs-6]
  "__xonsh__.subproc_captured_stdout(['ls'] + __xonsh__.pathsearch(foo, '[Ff]+i*LE', False, False) + ['-l'])"
# ---
# name: test_line_items[exprs.yml-globs-7]
  "print(__xonsh__.pathsearch(foo, '.*', True, False))"
# ---
# name: test_line_items[exprs.yml-globs-8]
  "__xonsh__.subproc_captured_stdout(['ls'] + __xonsh__.pathsearch(foo, '#[Ff]+i*LE', False, False) + ['-l'])"
# ---
# name: test_line_items[exprs.yml-globs-9]
  "print(__xonsh__.pathsearch(__xonsh__.regexsearch, '#.*', True, False))"
# ---
# name: test_line_items[exprs.yml-help-0]
  '__xonsh__.help(range)'
# ---
# name: test_line_items[exprs.yml-help-1]
  '__xonsh__.help(__xonsh__.help(range).index)'
# ---
# name: test_line_items[exprs.yml-help-2]
  '__xonsh__.superhelp(range)'
# ---
# name: test_line_items[exprs.yml-path-literals-0]
  "__xonsh__.path_literal('/foo')"
# ---
# name: test_line_items[exprs.yml-path-literals-10]
  "__xonsh__.path_literal(f'/foo{1 + 1}')"
# ---
# name: test_line_items[exprs.yml-path-literals-11]
  "__xonsh__.path_literal(f'/foo{1 + 1}')"
# ---
# name: test_line_items[exprs.yml-path-literals-12]
  "__xonsh__.path_literal(f'/foo{1 + 1}')"
# ---
# name: test_line_items[exprs.yml-path-literals-1]
  "__xonsh__.path_literal('/foo')"
# ---
# name: test_line_items[exprs.yml-path-literals-2]
  "__xonsh__.path_literal('/foo')"
# ---
# name: test_line_items[exprs.yml-path-literals-3]
  "__xonsh__.path_literal('/foo')"
# ---
# name: test_line_items[exprs.yml-path-literals-4]
  "__xonsh__.path_literal('/foo')"
# ---
# name: test_line_items[exprs.yml-path-literals-5]
  "__xonsh__.path_literal(f'/foo')"
# ---
# name: test_line_items[exprs.yml-path-literals-6]
  "__xonsh__.path_literal(f'/foo')"
# ---
# name: test_line_items[exprs.yml-path-literals-7]
  "__xonsh__.path_literal(f'/foo')"
# ---
# name: test_line_items[exprs.yml-path-literals-8]
  "__xonsh__.path_literal(f'/foo')"
# ---
# name: test_line_items[exprs.yml-path-literals-9]
  "__xonsh__.path_literal(f'/foo{1 + 1}')"
# ---
# name: test_line_items[exprs.yml-proc-injection-0]
  "__xonsh__.subproc_uncaptured([] + __xonsh__.subproc_captured_inject(['which', 'python']))"
# ---
# name: test_line_items[exprs.yml-proc-injection-1]
  "__xonsh__.subproc_uncaptured(['ls'] + __xonsh__.subproc_captured_inject(['dirname'] + __xonsh__.subproc_captured_inject(['which', 'python'])))"
# ---
# name: test_line_items[exprs.yml-proc-injection-2]
  "__xonsh__.subproc_captured_hiddenobject([] + __xonsh__.list_of_list_of_strs_outer_product(('a', __xonsh__.subproc_captured_inject(['echo', '1', '2']), 'b')))"
# ---
# name: test_line_items[exprs.yml-redirects-0]
  "__xonsh__.subproc_captured_object(['ls', ('>>', 'x.py')])"
# ---
# name: test_line_items[exprs.yml-redirects-1]
  "__xonsh__.subproc_captured_object(['ls'], '|', ['grep', 'wakka', ('>', 'x.py')])"
# ---
# name: test_line_items[exprs.yml-redirects-2]
  "__xonsh__.subproc_captured_object(['ls'], '|', ['grep', 'wakka', ('>>', 'x.py')])"
# ---
# name: test_line_items[exprs.yml-redirects-3]
  "__xonsh__.subproc_captured_stdout(['ls', ('>', 'x.py')])"
# ---
# name: test_line_items[exprs.yml-redirects-4]
  "__xonsh__.subproc_captured_stdout(['ls', ('>>', 'x.py')])"
# ---
# name: test_line_items[exprs.yml-redirects-5]
  "__xonsh__.subproc_captured_stdout(['ls'], '|', ['grep', 'wakka', ('>', 'x.py')])"
# ---
# name: test_line_items[exprs.yml-redirects-6]
  "__xonsh__.subproc_captured_stdout(['ls'], '|', ['grep', 'wakka', ('>>', 'x.py')])"
# ---
# name: test_line_items[exprs.yml-subproc-0]
  "__xonsh__.subproc_captured_stdout(['cmd', 'sub-cmd', '--opt'])"
# ---
# name: test_line_items[exprs.yml-subproc-10]
  "__xonsh__.subproc_captured_hiddenobject(['echo', ','])"
# ---
# name: test_line_items[exprs.yml-subproc-11]
  "__xonsh__.subproc_captured_hiddenobject(['echo', '1,2'])"
# ---
# name: test_line_items[exprs.yml-subproc-12]
  "__xonsh__.subproc_captured_object(['echo', '$foo'])"
# ---
# name: test_line_items[exprs.yml-subproc-13]
  "__xonsh__.subproc_captured_object(['echo', '$foo'])"
# ---
# name: test_line_items[exprs.yml-subproc-1]
  "__xonsh__.subproc_uncaptured(['cmd', 'sub-cmd', '--opt'])"
# ---
# name: test_line_items[exprs.yml-subproc-2]
  "__xonsh__.subproc_captured_hiddenobject(['cmd', 'sub-cmd', '--opt'])"
# ---
# name: test_line_items[exprs.yml-subproc-3]
  "__xonsh__.subproc_captured_object(['cmd', 'sub-cmd', '--opt'])"
# ---
# name: test_line_items[exprs.yml-subproc-4]
  "__xonsh__.subproc_captured_hiddenobject(['git', 'commit', '-am', 'wakka'])"
# ---
# name: test_line_items[exprs.yml-subproc-5]
  "__xonsh__.subproc_captured_hiddenobject(['git', 'commit', '-am', 'wakka jawaka'])"
# ---
# name: test_line_items[exprs.yml-subproc-6]
  "__xonsh__.subproc_captured_hiddenobject(['ls', 'wakka jawaka baraka'])"
# ---
# name: test_line_items[exprs.yml-subproc-7]
  "__xonsh__.subproc_uncaptured(['git', 'commit', '-am', 'wakka'])"
# ---
# name: test_line_items[exprs.yml-subproc-8]
  "__xonsh__.subproc_uncaptured(['git', 'commit', '-am', 'wakka jawaka'])"
# ---
# name: test_line_items[exprs.yml-subproc-9]
  "__xonsh__.subproc_uncaptured(['ls', 'wakka jawaka baraka'])"
# ---
# name: test_line_items[stmts.yml-assignments-0]
  "__xonsh__.env['WAKKA']"
# ---
# name: test_line_items[stmts.yml-assignments-6]
  "__xonsh__.env[None or 'WAKKA']"
# ---
# name: test_line_items[stmts.yml-assignments-7]
  "__xonsh__.env[__xonsh__.env['JAWAKA']]"
# ---
